lighthouse-core/lib/sd-validation/schema-validator.js

```diff
@@ -0,0 +1,131 @@
+const schemaStructure = require('./assets/schema-tree.json');
+const TYPE_KEYWORD = '@type';
+const SCHEMA_ORG_URL_REGEX = /https?:\/\/schema\.org\//;
+
+/**
+ * @param {string} uri
+ * @returns {string}
+ */
+function cleanName(uri) {
+  return uri.replace(SCHEMA_ORG_URL_REGEX, '');
+}
+
+/**
+ * @param {string} type
+ * @returns {Array<string>}
+ */
+function getPropsForType(type) {
+  const cleanType = cleanName(type);
+  const props = schemaStructure.properties
+    .filter(prop => prop.parent.includes(cleanType))
+    .map(prop => prop.name);
+  const foundType = findType(type);
+  if (!foundType) throw new Error(`Unable to get props for missing type "${type}"`);
+  const parentTypes = foundType.parent;
+
+  return parentTypes.reduce((allProps, type) => allProps.concat(getPropsForType(type)), props);
+}
+
+/**
+ * @param {string} type
+ * @returns {{name: string, parent: Array<string>}|undefined}
+ */
+function findType(type) {
+  const cleanType = cleanName(type);
+
+  return schemaStructure.types.find(typeObj => typeObj.name === cleanType);
+}
+
+/**
+ * Validates keys of given object based on its type(s). Returns an array of error messages.
+ *
+ * @param {string|Array<string>} typeOrTypes
+ * @param {Array<string>} keys
+ * @returns {Array<string>}
+ */
+function validateObjectKeys(typeOrTypes, keys) {
+  /** @type {Array<string>} */
+  let types = [];
+
+  if (typeof typeOrTypes === 'string') {
+    types = [typeOrTypes];
+  } else if (Array.isArray(typeOrTypes)) {
+    types = typeOrTypes;
+  } else {
+    return ['Unknown value type'];
+  }
+
+  const unknownTypes = types.filter(t => !findType(t));
+
+  if (unknownTypes.length) {
+    return unknownTypes
+      .filter(type => SCHEMA_ORG_URL_REGEX.test(type))
+      .map(type => `Unrecognized schema.org type ${type}`);
+  }
+
+  /** @type {Set<string>} */
+  const allKnownProps = new Set();
+
+  types.forEach(type => {
+    const knownProps = getPropsForType(type);
+
+    knownProps.forEach(key => allKnownProps.add(key));
+  });
+
+  const cleanKeys = keys
+    // Skip JSON-LD keywords (including invalid ones as they were already flagged in the json-ld validator)
+    .filter(key => key.indexOf('@') !== 0)
+    .map(key => cleanName(key));
+
+  return cleanKeys
+    // remove Schema.org input/output constraints http://schema.org/docs/actions.html#part-4
+    .map(key => key.replace(/-(input|output)$/, ''))
+    .filter(key => !allKnownProps.has(key))
+    .map(key => `Unexpected property "${key}"`);
+}
+
+/**
+ * @param {LH.StructuredData.ExpandedSchemaRepresentation|null} expandedObj Valid JSON-LD object in expanded form
+ * @return {Array<{path: string, message: string}>}
+ */
+module.exports = function validateSchemaOrg(expandedObj) {
+  /** @type {Array<{path: string, message: string}>} */
+  const errors = [];
+
+  if (expandedObj === null) {
+    return errors;
+  }
+
+  if (Array.isArray(expandedObj) && expandedObj.length === 1) {
```

Comment:
What if `.length` is > 1?  `walkObject()` does not explicitly support arrays as input, and will fail to traverse.

Not clear if it would necessarily be OK to have an array a valid schema.org objects, but it is OK to have a bunch of schema.org scripts on a page.

Reply:
According to [JSON-LD spec, section on grammar](https://www.w3.org/TR/json-ld/#h2_json-ld-grammar): "A JSON-LD document must be a single node object or an array whose elements are each node objects at the top level."

Reply:
that should be handled by the `walkObject` case below, added a test case to verify this 👍